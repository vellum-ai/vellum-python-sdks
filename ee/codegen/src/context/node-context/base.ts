import { VellumVariableType } from "vellum-ai/api";

import {
  VELLUM_WORKFLOW_BASE_NODES_MODULE_PATH,
  VELLUM_WORKFLOW_NODES_MODULE_PATH,
} from "src/constants";
import { WorkflowContext } from "src/context";
import { PortContext } from "src/context/port-context";
import { NodeOutputNotFoundError } from "src/generators/errors";
import { NodePort as NodePortType, WorkflowDataNode } from "src/types/vellum";
import { toValidPythonIdentifier } from "src/utils/casing";
import { getNodeLabel } from "src/utils/nodes";
import {
  doesModulePathStartWith,
  getGeneratedNodeDisplayModulePath,
  getGeneratedNodeModuleInfo,
} from "src/utils/paths";
import { hasErrorOutput } from "src/utils/typing";

export declare namespace BaseNodeContext {
  interface Args<T extends WorkflowDataNode> {
    workflowContext: WorkflowContext;
    nodeData: T;
    importOrder: readonly number[];
  }
}

export abstract class BaseNodeContext<T extends WorkflowDataNode> {
  public readonly workflowContext: WorkflowContext;
  public readonly importOrder: readonly number[];
  public readonly nodeModulePath: string[];
  public readonly nodeModuleName: string;
  public readonly nodeFileName: string;
  public readonly nodeClassName: string;
  public readonly nodeDisplayModulePath: string[];
  public readonly nodeDisplayModuleName: string;
  public readonly nodeDisplayFileName: string;
  public readonly nodeDisplayClassName: string;

  public nodeData: T;
  public abstract readonly baseNodeClassName: string;
  public abstract readonly baseNodeDisplayClassName: string;

  protected isCore = false;
  public readonly baseNodeClassModulePath: readonly string[];
  public readonly baseNodeDisplayClassModulePath: readonly string[];

  private nodeOutputNamesById: Record<string, string> | undefined;
  public readonly portContextsById: Map<string, PortContext>;
  public readonly defaultPortContext: PortContext | undefined;

  private readonly nodePortNames: Set<string> = new Set();

  constructor(args: BaseNodeContext.Args<T>) {
    this.workflowContext = args.workflowContext;
    this.nodeData = args.nodeData;
    this.importOrder = args.importOrder;

    if (args.nodeData.type === "GENERIC") {
      if (
        !doesModulePathStartWith(
          args.nodeData.base?.module,
          VELLUM_WORKFLOW_BASE_NODES_MODULE_PATH
        )
      ) {
        this.baseNodeClassModulePath = args.nodeData.base.module;
      } else {
        this.baseNodeClassModulePath = VELLUM_WORKFLOW_NODES_MODULE_PATH;
      }
    } else if (
      !doesModulePathStartWith(
        args.nodeData.base?.module,
        VELLUM_WORKFLOW_NODES_MODULE_PATH
      )
    ) {
      this.baseNodeClassModulePath =
        args.nodeData.base?.module ??
        this.workflowContext.sdkModulePathNames.DISPLAYABLE_NODES_MODULE_PATH;
    } else {
      this.baseNodeClassModulePath =
        this.workflowContext.sdkModulePathNames.DISPLAYABLE_NODES_MODULE_PATH;
    }
    this.baseNodeDisplayClassModulePath =
      this.workflowContext.sdkModulePathNames.NODE_DISPLAY_MODULE_PATH;

    const { moduleName, nodeClassName, modulePath } =
      getGeneratedNodeModuleInfo({
        workflowContext: args.workflowContext,
        nodeDefinition: args.nodeData.definition,
        nodeLabel: this.getNodeLabel(),
      });
    this.nodeModuleName = moduleName;
    this.nodeClassName = nodeClassName;
    this.nodeModulePath = modulePath;

    this.workflowContext.addUsedClassName(nodeClassName);
    this.nodeFileName = `${this.nodeModuleName}.py`;

    this.nodeDisplayModuleName = this.nodeModuleName;
    this.nodeDisplayFileName = `${this.nodeDisplayModuleName}.py`;
    this.nodeDisplayClassName = `${this.nodeClassName}Display`;
    this.nodeDisplayModulePath = getGeneratedNodeDisplayModulePath(
      args.workflowContext,
      this.nodeDisplayModuleName
    );

    const portContexts = this.nodeData.ports
      ? this.createPortContextsFromNodePortAttribute(this.nodeData.ports)
      : this.createPortContexts();
    portContexts.forEach((portContext) => {
      this.workflowContext.addPortContext(portContext);
    });

    this.portContextsById = new Map(
      portContexts.map((portContext) => [portContext.portId, portContext])
    );

    this.defaultPortContext = portContexts.find(
      (portContext) => portContext.isDefault
    );
  }

  protected abstract getNodeOutputNamesById(): Record<string, string>;
  protected abstract getNodeOutputTypesById(): Record<
    string,
    VellumVariableType
  >;
  protected abstract createPortContexts(): PortContext[];

  protected createPortContextsFromNodePortAttribute(
    nodePorts: NodePortType[]
  ): PortContext[] {
    return nodePorts.map(
      (port) =>
        new PortContext({
          workflowContext: this.workflowContext,
          nodeContext: this,
          portId: port.id,
          portName: port.name,
          isDefault: port.type === "DEFAULT",
        })
    );
  }

  public getNodeLabel(): string {
    return getNodeLabel(this.nodeData);
  }

  public getNodeOutputNameById(outputId: string): string | undefined {
    // Lazily load node output names
    if (!this.nodeOutputNamesById) {
      this.nodeOutputNamesById = this.getNodeOutputNamesById();
    }

    const nodeOutputName = this.nodeOutputNamesById[outputId];
    if (!nodeOutputName) {
      // This handles an edge case where the node references a non-existent subworkflow deployment
      if (
        this.nodeData.type === "SUBWORKFLOW" &&
        this.nodeData.data.variant === "DEPLOYMENT"
      ) {
        this.workflowContext.addError(
          new NodeOutputNotFoundError(
            `Could not find Subworkflow Deployment Output with id ${outputId}`,
            "WARNING"
          )
        );
        return;
      }

      this.workflowContext.addError(
        new NodeOutputNotFoundError(
          `Failed to find ${this.nodeClassName} Output with id '${outputId}'`,
          "WARNING"
        )
      );
      return;
    }

    return toValidPythonIdentifier(nodeOutputName, "output");
  }

  public getNodeOutputTypeById(
    outputId: string
  ): VellumVariableType | undefined {
    const nodeOutputTypesById = this.getNodeOutputTypesById();

    return nodeOutputTypesById[outputId];
  }

  public isPortNameUsed(portName: string): boolean {
    return this.nodePortNames.has(portName);
  }

  public addUsedPortName(portName: string): void {
    this.nodePortNames.add(portName);
  }

  /**
   * Extend this class to perform any additional property generation asynchronously
   * after the node context has been created.
   */
  public buildProperties(): Promise<void> {
    return Promise.resolve();
  }

  public isImportedBefore(
    nodeContext: BaseNodeContext<WorkflowDataNode>
  ): boolean {
    for (
      let i = 0;
      i < Math.max(this.importOrder.length, nodeContext.importOrder.length);
      i++
    ) {
      const thisImportOrder = this.importOrder[i];
      const otherImportOrder = nodeContext.importOrder[i];
      if (thisImportOrder === undefined && otherImportOrder === undefined) {
        return false;
      }

      if (thisImportOrder === undefined) {
        return true;
      }

      if (otherImportOrder === undefined) {
        return false;
      }

      if (thisImportOrder < otherImportOrder) {
        return true;
      }

      if (thisImportOrder > otherImportOrder) {
        return false;
      }
    }
    return false;
  }

  protected getErrorOutputId(): string | undefined {
    const TRY_NODE_NAME = "TryNode";
    const tryNodes = this.nodeData.adornments?.filter((adornment) => {
      return adornment.base.name === TRY_NODE_NAME;
    });

    const tryNode = tryNodes?.at(-1);

    let fallbackErrorOutputId: string | undefined = undefined;
    if (hasErrorOutput(this.nodeData)) {
      fallbackErrorOutputId = this.nodeData.data.errorOutputId;
    }

    return tryNode ? tryNode.id : fallbackErrorOutputId;
  }
}
