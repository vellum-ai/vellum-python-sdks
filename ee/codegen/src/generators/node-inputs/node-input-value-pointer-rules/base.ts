import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";

import { WorkflowContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import {
  IterableConfig,
  NodeInputValuePointerRule as NodeInputValuePointerRuleType,
  WorkflowDataNode,
} from "src/types/vellum";

export declare namespace BaseNodeInputValuePointerRule {
  export interface Args<T extends NodeInputValuePointerRuleType> {
    nodeContext: BaseNodeContext<WorkflowDataNode>;
    nodeInputValuePointerRule: T;
    iterableConfig?: IterableConfig;
  }
}

export abstract class BaseNodeInputValuePointerRule<
  T extends NodeInputValuePointerRuleType
> extends AstNode {
  private readonly nodeContext: BaseNodeContext<WorkflowDataNode>;
  public readonly workflowContext: WorkflowContext;
  public readonly nodeInputValuePointerRule: T;
  public readonly iterableConfig?: IterableConfig;
  private astNode: AstNode | undefined;

  constructor({
    nodeContext,
    nodeInputValuePointerRule,
    iterableConfig,
  }: BaseNodeInputValuePointerRule.Args<T>) {
    super();
    this.nodeContext = nodeContext;
    this.workflowContext = nodeContext.workflowContext;
    this.iterableConfig = iterableConfig;
    this.nodeInputValuePointerRule = nodeInputValuePointerRule;

    this.astNode = this.getAstNode();
    if (this.astNode) {
      this.inheritReferences(this.astNode);
    }
  }

  public getReferencedNodeContext():
    | BaseNodeContext<WorkflowDataNode>
    | undefined {
    return undefined;
  }

  abstract getAstNode(): AstNode | undefined;

  public write(writer: Writer): void {
    if (this.astNode) {
      this.astNode.write(writer);
    }
  }
}
