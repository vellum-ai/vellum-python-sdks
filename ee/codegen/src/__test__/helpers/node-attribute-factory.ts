import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
