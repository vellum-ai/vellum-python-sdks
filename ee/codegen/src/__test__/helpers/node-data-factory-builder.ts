import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
