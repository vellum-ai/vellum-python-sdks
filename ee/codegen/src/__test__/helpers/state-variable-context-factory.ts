import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
