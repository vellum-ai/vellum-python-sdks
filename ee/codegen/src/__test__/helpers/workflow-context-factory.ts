import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
