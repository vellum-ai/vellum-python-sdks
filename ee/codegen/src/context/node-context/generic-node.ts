import { BaseNodeContext } from "src/context/node-context/base";
import { PortContext } from "src/context/port-context";
import { GenericNode as GenericNodeType } from "src/types/vellum";

export class GenericNodeContext extends BaseNodeContext<GenericNodeType> {
  baseNodeClassName = "BaseNode";
  baseNodeDisplayClassName = "BaseNodeDisplay";

  getNodeOutputNamesById(): Record<string, string> {
    return Object.fromEntries(
      this.nodeData.outputs.map((output) => [output.id, output.name])
    );
  }

  createPortContexts(): PortContext[] {
    return this.nodeData.ports.map(
      (port) =>
        new PortContext({
          workflowContext: this.workflowContext,
          nodeContext: this,
          portId: port.id,
          portName: port.name,
          isDefault: port.type === "DEFAULT",
        })
    );
  }
}
