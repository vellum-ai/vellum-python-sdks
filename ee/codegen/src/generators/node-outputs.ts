import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";

import { OUTPUTS_CLASS_NAME } from "src/constants";
import { WorkflowContext } from "src/context";
import { GenericNodeContext } from "src/context/node-context/generic-node";
import { WorkflowValueDescriptor } from "src/generators/workflow-value-descriptor";
import { NodeOutput as NodeOutputType } from "src/types/vellum";
import { getVellumVariablePrimitiveType } from "src/utils/vellum-variables";

export declare namespace NodeOutputs {
  export interface Args {
    nodeOutputs: NodeOutputType[];
    nodeContext: GenericNodeContext;
    workflowContext: WorkflowContext;
  }
}

export class NodeOutputs extends AstNode {
  private astNode: AstNode;

  public constructor(args: NodeOutputs.Args) {
    super();

    this.astNode = this.constructNodeOutputs(
      args.nodeOutputs,
      args.nodeContext,
      args.workflowContext
    );
  }

  private constructNodeOutputs(
    nodeOutputs: NodeOutputType[],
    nodeContext: GenericNodeContext,
    workflowContext: WorkflowContext
  ): AstNode {
    const baseNodeClassNameAlias =
      nodeContext.baseNodeClassName === nodeContext.nodeClassName
        ? `Base${nodeContext.baseNodeClassName}`
        : undefined;

    const clazz = python.class_({
      name: OUTPUTS_CLASS_NAME,
      extends_: [
        python.reference({
          name: nodeContext.baseNodeClassName,
          modulePath: nodeContext.baseNodeClassModulePath,
          alias: baseNodeClassNameAlias,
          attribute: [OUTPUTS_CLASS_NAME],
        }),
      ],
    });

    nodeOutputs.forEach((output) => {
      const type = getVellumVariablePrimitiveType(output.type);
      const field = output.value
        ? python.field({
            name: output.name,
            initializer: new WorkflowValueDescriptor({
              workflowValueDescriptor: output.value,
              workflowContext,
              nodeContext,
            }),
          })
        : python.field({
            name: output.name,
            type: type,
          });
      clazz.addField(field);
    });

    return clazz;
  }

  write(writer: Writer): void {
    if (this.astNode) {
      this.astNode.write(writer);
    }
  }
}
