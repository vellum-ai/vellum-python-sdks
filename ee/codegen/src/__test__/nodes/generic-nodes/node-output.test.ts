import { Writer } from "@fern-api/python-ast/core/Writer";
import { beforeEach, describe, expect, it } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { inputVariableContextFactory } from "src/__test__/helpers/input-variable-context-factory";
import { genericNodeFactory } from "src/__test__/helpers/node-data-factories";
import { createNodeContext, WorkflowContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import { GenericNodeContext } from "src/context/node-context/generic-node";
import { NodeOutputs } from "src/generators/node-outputs";
import {
  NodeOutput as NodeOutputType,
  WorkflowDataNode,
} from "src/types/vellum";

describe("NodeOutput", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let nodeOutput: NodeOutputs;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
    writer = new Writer();
  });

  describe("basic", () => {
    beforeEach(async () => {
      const nodeOutputData: NodeOutputType[] = [
        {
          id: "8c3c9aff-e1d5-49f4-af75-3ec2fcbb4af2",
          name: "output",
          type: "NUMBER",
        },
      ];
      const nodeData = genericNodeFactory({
        name: "AnnotatedOutputGenericNode",
        nodeOutputs: nodeOutputData,
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as GenericNodeContext;

      nodeOutput = new NodeOutputs({
        nodeOutputs: nodeOutputData,
        nodeContext,
        workflowContext,
      });
    });

    it("generates output class with number type", async () => {
      nodeOutput.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("workflow input output", () => {
    beforeEach(async () => {
      const nodeOutputData: NodeOutputType[] = [
        {
          id: "2c4a85c0-b017-4cea-a261-e8e8498570c9",
          name: "output",
          type: "STRING",
          value: {
            type: "WORKFLOW_INPUT",
            inputVariableId: "some-id",
          },
        },
      ];
      const nodeData = genericNodeFactory({
        name: "WorkflowInputGenericNode",
        nodeOutputs: nodeOutputData,
      });

      workflowContext.addInputVariableContext(
        inputVariableContextFactory({
          inputVariableData: {
            id: "some-id",
            key: "count",
            type: "NUMBER",
          },
          workflowContext,
        })
      );

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as GenericNodeContext;

      nodeOutput = new NodeOutputs({
        nodeOutputs: nodeOutputData,
        nodeContext,
        workflowContext,
      });
    });

    it("generates correct output class with workflow input reference", async () => {
      nodeOutput.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("node output reference", () => {
    beforeEach(async () => {
      vi.spyOn(workflowContext, "getNodeContext").mockReturnValue({
        nodeClassName: "TestNode",
        path: ["nodes", "test-node-path"],
        getNodeOutputNameById: vi.fn().mockReturnValue("my_output"),
      } as unknown as BaseNodeContext<WorkflowDataNode>);
      const nodeOutputData: NodeOutputType[] = [
        {
          id: "db010db3-7076-4df9-ae1b-069caa16fa20",
          name: "output",
          type: "STRING",
          value: {
            type: "NODE_OUTPUT",
            nodeId: "some-node-id",
            nodeOutputId: "some-output-id",
          },
        },
      ];

      const nodeData = genericNodeFactory({
        name: "GenericNodeReferencingOutput",
        nodeOutputs: nodeOutputData,
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as GenericNodeContext;

      nodeOutput = new NodeOutputs({
        nodeOutputs: nodeOutputData,
        nodeContext,
        workflowContext,
      });
    });

    it("generates correct output class with node output reference", async () => {
      nodeOutput.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
