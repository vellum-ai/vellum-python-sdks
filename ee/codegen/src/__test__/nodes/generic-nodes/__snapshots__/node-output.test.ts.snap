// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NodeOutput > basic > generates output class with number type 1`] = `
"class Outputs(BaseNode.Outputs):
    output: Union[float, int]
"
`;

exports[`NodeOutput > node output reference > generates correct output class with node output reference 1`] = `
"class Outputs(BaseNode.Outputs):
    output = TestNode.Outputs.my_output
"
`;

exports[`NodeOutput > workflow input output > generates correct output class with workflow input reference 1`] = `
"class Outputs(BaseNode.Outputs):
    output = Inputs.count
"
`;
